### PR TITLE
Add support for route parameters in fragments

### PR DESCRIPTION
**Pull request by BlakeWilliams**
Currently Voltron supports page fragments in routes, but not fragment paths. The current approach is to always turn those named fragments into query parameters and pass them into the route.

So far this has two downsides:
* There can't be a query parameter passed to the route that has the same name as a route parameter
* Rails expects these parameters to be route parameters, not query parameters. This breaks the Rails behavior of relying on route parameters to be automatically used in path helpers.

This addresses the issue by automatically converting the route parameters from the route into route parameters in the fragment. When a new route is added, we validate that the route parameters are also accepted by each fragment unless they contain the "legacy" metadata field.

### CODE REVIEW

File: pkg/fragment/definition.go [hunk 1]
```diff
@@ -43,19 +62,44 @@ func WithTimingLabel(timingLabel string) DefinitionOption {
...
 	// This is already parsed before constructing the url in server.go, so we ignore errors
 	targetUrl, _ := url.Parse(d.Url)
 	targetUrl.RawQuery = parameters.Encode()
 
-	return targetUrl.String()
+	return targetUrl
 }
 
-func (d *Definition) IntoRequestable(params url.Values) multiplexer.Requestable {
+func (d *Definition) Requestable(target *url.URL, parameters map[string]string, query url.Values) (*Request, error) {
```
Comment (iancanderson) [inline pkg/fragment/definition.go:77]:
```suggestion
func (d *Definition) Requestable(target *url.URL, pathParams map[string]string, query url.Values) (*Request, error) {
```

Other functions in this file refer to the query params as `parameters`. Hoping to disambiguate

File: pkg/fragment/definition.go [hunk 2]
```diff
@@ -43,19 +62,44 @@ func WithTimingLabel(timingLabel string) DefinitionOption {
...
+	request := *target // clone the url
+
+	var path strings.Builder
+
+	for _, part := range d.routeParts {
+		path.WriteByte('/')
+
+		if strings.HasPrefix(part, ":") {
+			if replacement, ok := parameters[part]; ok {
```
Comment (iancanderson) [inline pkg/fragment/definition.go:86]:
```suggestion
			if replacement, ok := pathParams[part]; ok {
```

**Review by iancanderson — APPROVED** [review]
Nice!

File: pkg/fragment/definition.go (hunk 1 above)
Comment (BlakeWilliams) [reply pkg/fragment/definition.go:77]: That's a lot more clear, thanks!

File: pkg/fragment/definition.go [hunk 3]
```diff
@@ -12,25 +12,44 @@ type Collection = []*Definition
...
-	Url         string
-	Metadata    map[string]string `json:"metadata"`
-	TimingLabel string            `json:"timingLabel"`
+	Path             string
+	routeParts       []string
+	dynamicParts     []string
+	Url              string
+	Metadata         map[string]string
+	TimingLabel      string
+	IgnoreValidation bool `json:"ignoreValidation"`
```
Comment (kevinsawicki) [inline pkg/fragment/definition.go:21]: Is this json key needed? Looks like the default name is used.

File: pkg/fragment/definition.go [hunk 4]
```diff
@@ -43,19 +62,44 @@ func WithTimingLabel(timingLabel string) DefinitionOption {
...
+	var path strings.Builder
+
+	for _, part := range d.routeParts {
+		path.WriteByte('/')
+
+		if strings.HasPrefix(part, ":") {
+			if replacement, ok := pathParams[part]; ok {
+				path.WriteString(replacement)
+			} else {
+				return nil, fmt.Errorf("no url replacement found for %s in %s", part, d.Path)
```
Comment (kevinsawicki) [inline pkg/fragment/definition.go:89]: What do you think about testing this from `pkg/fragment/definition_test.go` ? Or was it covered in tests in other packages?

File: pkg/routeimporter/config.go [hunk 5]
```diff
@@ -6,16 +6,44 @@ import (
...
 	return nil
 }
+
+func createFragment(template *fragment.Definition) *fragment.Definition {
+	fragment := fragment.Define(template.Path, fragment.WithMetadata(template.Metadata))
+	fragment.TimingLabel = template.TimingLabel
+
+	if template.IgnoreValidation {
+		fragment.IgnoreValidation = true
+	}
```
Comment (kevinsawicki) [inline pkg/routeimporter/config.go:46]: Would seem okay to pass this through the same as `TimingLabel`

```suggestion
	fragment.IgnoreValidation = template.IgnoreValidation
```

File: server_test.go [hunk 6]
```diff
@@ -454,6 +502,34 @@ func startTargetServer() *httptest.Server {
...
+			w.WriteHeader(http.StatusOK)
+			w.Write([]byte("<body>"))
+		} else if strings.HasPrefix(r.URL.Path, "/body/") {
+			w.WriteHeader(http.StatusOK)
+			w.Write([]byte(fmt.Sprintf("hello %s", name)))
+		} else if strings.HasPrefix(r.URL.Path, "/footer/") {
+			w.WriteHeader(http.StatusOK)
+			w.Write([]byte("</body>"))
+		} else {
+			fmt.Println(parts)
```
Comment (kevinsawicki) [inline server_test.go:523]:
```suggestion
```

**Review by kevinsawicki — APPROVED** [review]
Looks good to me, left a few minor comments

File: pkg/fragment/definition.go (hunk 4 above)
Comment (BlakeWilliams) [reply pkg/fragment/definition.go]: I like it, added in https://github.com/BlakeWilliams/viewproxy/pull/71/commits/d888a6c673e546f8fb4522ba4a983adbf44a6ae7

**Pull request merged**